src/eei.cpp

```diff
@@ -90,6 +90,80 @@ Literal EthereumInterface::callImport(Import *import, LiteralList& arguments) {
       return Literal();
     }
 
+    if (
+      import->base == Name("call") ||
+      import->base == Name("callCode") ||
+      import->base == Name("callDelegate") ||
+      import->base == Name("callStatic")
+    ) {
+      int64_t gas = arguments[0].geti64();
+      uint32_t addressOffset = arguments[1].geti32();
+      uint32_t valueOffset;
+      uint32_t dataOffset;
+      uint32_t dataLength;
+      uint32_t resultOffset;
+      uint32_t resultLength;
+
+      bool isValuePresent = import->base == Name("call") || import->base == Name("callCode");
+      evm_result call_result;
+      evm_message call_message;
+
+      heraAssert((msg.flags & ~EVM_STATIC) == 0, "Unknown flags not supported.");
+
+      if (isValuePresent) {
+      	valueOffset = arguments[2].geti32();
+	dataOffset = arguments[3].geti32();
+	dataLength = arguments[4].geti32();
+	resultOffset = arguments[5].geti32();
+	resultLength = arguments[6].geti32();
+
+	call_message.sender = msg.address;
+	call_message.value = loadUint128(valueOffset);
+	call_message.kind = (import->base == Name("callCode")) ? EVM_CALLCODE : EVM_CALL;
+	call_message.flags = 0;
+      } else {
+      	dataOffset = arguments[2].geti32();
+	dataLength = arguments[3].geti32();
+	resultOffset = arguments[4].geti32();
+	resultLength = arguments[5].geti32();
+
+	if (import->base == Name("callDelegate")) {
+	  call_message.sender = msg.sender;
+	  call_message.value = msg.value;
+	  call_message.kind = EVM_DELEGATECALL;
+	  call_message.flags = 0;
+	} else if (import->base == Name("callStatic")) {
+	  call_message.sender = msg.address;
+	  call_message.value = { };
+	  call_message.kind = EVM_CALL;
+	  call_message.flags |= EVM_STATIC;
+	}
+      }
+
+      vector<uint8_t> input_data(dataLength);
+
+      call_message.address = loadUint160(addressOffset);
+      loadMemory(dataOffset, input_data, dataLength);
+      call_message.input = input_data.data();
+      call_message.input_size = dataLength;
+      call_message.code_hash = { };
+      call_message.gas = gas;
+      call_message.depth = msg.depth + 1;
+
+      context->fn_table->call(&call_result, context, &call_message);
+
+      if (call_result.output_data) {
+      	 vector<uint8_t> result(call_result.output_data, call_result.output_data + call_result.output_size);
+	 result.resize(resultLength);
+	 storeMemory(result, 0, resultOffset, resultLength);
+      }
+
+      if (call_result.release)
+        call_result.release(&call_result);
+
+      return Literal((call_result.status_code == EVM_SUCCESS) ? 1 : 0);
```

Comment:
@axic would this not be the other way around? The spec states that this should return 1 when the VM traps, and 0 otherwise.

Reply:
Oh probably we fixed it in eWASM. EVM has a bool. In that case eWASM should return states properly: 0 (success), 1 (fail), 2 (revert)